examples/text/plot_document_clustering.py

```diff
@@ -3,284 +3,328 @@
-    print("Performing dimensionality reduction using LSA")
-    t0 = time()
-    # Since LSA/SVD results are not normalized,
-    # we redo the normalization to improve the k-means result.
-    svd = TruncatedSVD(opts.n_components)
-    normalizer = Normalizer(copy=False)
-    lsa = make_pipeline(svd, normalizer)
+from sklearn.feature_extraction.text import HashingVectorizer
 
-    X = lsa.fit_transform(X)
+hasher = HashingVectorizer(
+    stop_words="english",
+    alternate_sign=False,
+    norm=None,
+)
 
-    print("done in %fs" % (time() - t0))
+t0 = time()
+X = hasher.fit_transform(dataset.data)
+print(f"vectorizing done in {time() - t0:.3f} s")
+print(f"n_samples: {X.shape[0]}, n_features: {X.shape[1]}")
 
-    explained_variance = svd.explained_variance_ratio_.sum()
-    print(
-        "Explained variance of the SVD step: {}%".format(int(explained_variance * 100))
-    )
+fit_and_evaluate(kmeans, X, name="kmeans with\nsimple hashing")
 
-    print()
+# %%
+# When IDF weighting is needed it can be added by pipelining the
+# :class:`~sklearn.feature_extraction.text.HashingVectorizer` output to a
+# :class:`~sklearn.feature_extraction.text.TfidfTransformer` instance.
 
+from sklearn.feature_extraction.text import TfidfTransformer
 
-# %%
-# Clustering
-# ----------
-
-if opts.minibatch:
-    km = MiniBatchKMeans(
-        n_clusters=true_k,
-        init="k-means++",
-        n_init=1,
-        init_size=1000,
-        batch_size=1000,
-        verbose=opts.verbose,
-    )
-else:
-    km = KMeans(
-        n_clusters=true_k,
-        init="k-means++",
-        max_iter=100,
-        n_init=1,
-        verbose=opts.verbose,
-    )
+vectorizer = make_pipeline(hasher, TfidfTransformer())
 
-print("Clustering sparse data with %s" % km)
 t0 = time()
-km.fit(X)
-print("done in %0.3fs" % (time() - t0))
-print()
+X = vectorizer.fit_transform(dataset.data)
+print(f"vectorizing done in {time() - t0:.3f} s")
 
+fit_and_evaluate(kmeans, X, name="kmeans with\nTfidf-scaled hashing")
 
 # %%
-# Performance metrics
-# -------------------
-
-print("Homogeneity: %0.3f" % metrics.homogeneity_score(labels, km.labels_))
-print("Completeness: %0.3f" % metrics.completeness_score(labels, km.labels_))
-print("V-measure: %0.3f" % metrics.v_measure_score(labels, km.labels_))
-print("Adjusted Rand-Index: %.3f" % metrics.adjusted_rand_score(labels, km.labels_))
-print(
-    "Silhouette Coefficient: %0.3f"
-    % metrics.silhouette_score(X, km.labels_, sample_size=1000)
-)
+# Plot unsupervised evaluation metrics
+# ====================================
+
+import pandas as pd
 
-print()
+df = pd.DataFrame(scores).set_index("estimator")
 
+ax = df.plot.barh(figsize=(12, 6))
+ax.set_ylabel("")
+_ = ax.set_xlabel("Clustering scores")
 
 # %%
+# It can be noticed that :class:`~sklearn.cluster.KMeans` (and
+# :class:`~sklearn.cluster.MiniBatchKMeans`) are very sensitive to feature
+# scaling and that in this case the IDF weighting helps improve the quality of
```

Comment:
I think that it would help here to refer to the exact labels on the graph (hasher vs. TFIDF scaled hasher)